### ISSUE TITLE
Not started engine if async open_spider

### ISSUE THREAD

**Issue by tonal** [scrapy/scrapy#4887]
[Description of the issue]

In 2.4.0 version
If the pipelines or extensions has an **async** signal handler **open_spider**, the engine will not start.

Bat 2.3.0 version work

### Steps to Reproduce

1. pip install Scrapy uvloop
2. file quotes_spider.py:
```python
# -*- coding: utf-8 -*-

from scrapy import signals, Spider


class QuotesSpider(Spider):
    name = 'quotes'
    start_urls = [
        'http://quotes.toscrape.com/tag/humor/',
    ]

    custom_settings = dict(
        TWISTED_REACTOR='twisted.internet.asyncioreactor.AsyncioSelectorReactor',
        ASYNCIO_EVENT_LOOP='uvloop.Loop',
        EXTENSIONS={'quotes_spider.AsyncOpenExt': 10},
    )

    def parse(self, response, **kwargs):
        for quote in response.css('div.quote'):
            yield {
                'author': quote.xpath('span/small/text()').get(),
                'text': quote.css('span.text::text').get(),
            }

        next_page = response.css('li.next a::attr("href")').get()
        if next_page is not None:
            yield response.follow(next_page, self.parse)


class AsyncOpenExt:

    @classmethod
    def from_crawler(cls, crawler):
        ext = cls()
        crawler.signals.connect(ext.spider_opened, signal=signals.spider_opened)
        return ext

    async def spider_opened(self, spider:Spider, signal=None, sender=None):
        pass
```
3. Start scrapy:
```bash
$ scrapy runspider quotes_spider.py -o quotes.jl -s TWISTED_REACTOR=twisted.internet.asyncioreactor.AsyncioSelectorReactor -s ASYNCIO_EVENT_LOOP=uvloop.Loop
```
**Expected behavior:**
```
2020-11-14 14:39:49 [scrapy.utils.log] INFO: Scrapy 2.3.0 started (bot: scrapybot)
2020-11-14 14:39:49 [scrapy.utils.log] INFO: Versions: lxml 4.6.1.0, libxml2 2.9.10, cssselect 1.1.0, parsel 1.6.0, w3lib 1.22.0, Twisted 20.3.0, Python 3.8.6 (default, Sep 28 2020, 10:21:29) - [GCC 9.3.0], pyOpenSSL 19.1.0 (OpenSSL 1.1.1h  22 Sep 2020), cryptography 3.2.1, Platform Linux-5.4.0-53-generic-x86_64-with-glibc2.2.5
2020-11-14 14:39:49 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor
...
2020-11-14 14:39:49 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2020-11-14 14:39:49 [scrapy.core.engine] INFO: Spider opened
2020-11-14 14:39:49 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2020-11-14 14:39:49 [scrapy.extensions.telnet] INFO: Telnet console listening on 127.0.0.1:6023
2020-11-14 14:39:50 [scrapy.core.engine] DEBUG: Crawled (200) <GET http://quotes.toscrape.com/tag/humor/> (referer: None)
...
2020-11-14 14:39:50 [scrapy.core.engine] INFO: Closing spider (finished)
2020-11-14 14:39:50 [scrapy.extensions.feedexport] INFO: Stored jl feed (12 items) in: quotes.jl
2020-11-14 14:39:50 [scrapy.statscollectors] INFO: Dumping Scrapy stats:
{'downloader/request_bytes': 513,
...
 'finish_reason': 'finished',
...
 'start_time': datetime.datetime(2020, 11, 14, 7, 39, 49, 965142)}
2020-11-14 14:39:50 [scrapy.core.engine] INFO: Spider closed (finished)
```
**Actual behavior:**
```
2020-11-14 14:41:39 [scrapy.utils.log] INFO: Scrapy 2.4.0 started (bot: scrapybot)
2020-11-14 14:41:39 [scrapy.utils.log] INFO: Versions: lxml 4.6.1.0, libxml2 2.9.10, cssselect 1.1.0, parsel 1.6.0, w3lib 1.22.0, Twisted 20.3.0, Python 3.8.6 (default, Sep 28 2020, 10:21:29) - [GCC 9.3.0], pyOpenSSL 19.1.0 (OpenSSL 1.1.1h  22 Sep 2020), cryptography 3.2.1, Platform Linux-5.4.0-53-generic-x86_64-with-glibc2.2.5
2020-11-14 14:41:39 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor
2020-11-14 14:41:39 [scrapy.utils.log] DEBUG: Using asyncio event loop: uvloop.Loop
...
2020-11-14 14:41:40 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2020-11-14 14:41:40 [scrapy.core.engine] INFO: Spider opened
2020-11-14 14:41:40 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2020-11-14 14:41:40 [asyncio] DEBUG: Using selector: EpollSelector
2020-11-14 14:42:40 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
^C2020-11-14 14:42:51 [scrapy.crawler] INFO: Received SIGINT, shutting down gracefully. Send again to force 
Unhandled error in Deferred:
2020-11-14 14:42:51 [twisted] CRITICAL: Unhandled error in Deferred:

Traceback (most recent call last):
  File "/home/tonal/.pyenv/versions/3.8.6/envs/err-scrapy-async/lib/python3.8/site-packages/twisted/internet/defer.py", line 1613, in unwindGenerator
    return _cancellableInlineCallbacks(gen)
  File "/home/tonal/.pyenv/versions/3.8.6/envs/err-scrapy-async/lib/python3.8/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInlineCallbacks
    _inlineCallbacks(None, g, status)
  File "/home/tonal/.pyenv/versions/3.8.6/envs/err-scrapy-async/lib/python3.8/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/home/tonal/.pyenv/versions/3.8.6/envs/err-scrapy-async/lib/python3.8/site-packages/scrapy/crawler.py", line 109, in stop
    yield defer.maybeDeferred(self.engine.stop)
--- <exception caught here> ---
  File "/home/tonal/.pyenv/versions/3.8.6/envs/err-scrapy-async/lib/python3.8/site-packages/scrapy/crawler.py", line 109, in stop
    yield defer.maybeDeferred(self.engine.stop)
  File "/home/tonal/.pyenv/versions/3.8.6/envs/err-scrapy-async/lib/python3.8/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/tonal/.pyenv/versions/3.8.6/envs/err-scrapy-async/lib/python3.8/site-packages/scrapy/core/engine.py", line 87, in stop
    raise RuntimeError("Engine not running")
builtins.RuntimeError: Engine not running

2020-11-14 14:42:51 [twisted] CRITICAL: 
Traceback (most recent call last):
  File "/home/tonal/.pyenv/versions/3.8.6/envs/err-scrapy-async/lib/python3.8/site-packages/scrapy/crawler.py", line 109, in stop
    yield defer.maybeDeferred(self.engine.stop)
  File "/home/tonal/.pyenv/versions/3.8.6/envs/err-scrapy-async/lib/python3.8/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/tonal/.pyenv/versions/3.8.6/envs/err-scrapy-async/lib/python3.8/site-packages/scrapy/core/engine.py", line 87, in stop
    raise RuntimeError("Engine not running")
RuntimeError: Engine not running
```

**Reproduces how often:**
Always after upgrade to 2.4.0

### Versions

```
$ scrapy version --verbose
Scrapy       : 2.4.0
lxml         : 4.6.1.0
libxml2      : 2.9.10
cssselect    : 1.1.0
parsel       : 1.6.0
w3lib        : 1.22.0
Twisted      : 20.3.0
Python       : 3.8.6 (default, Sep 28 2020, 10:21:29) - [GCC 9.3.0]
pyOpenSSL    : 19.1.0 (OpenSSL 1.1.1h  22 Sep 2020)
cryptography : 3.2.1
Platform     : Linux-5.4.0-53-generic-x86_64-with-glibc2.2.5
```

### Additional context

Operating System: KDE neon 5.20
KDE Plasma Version: 5.20.3
KDE Frameworks Version: 5.75.0
Qt Version: 5.15.1
Kernel Version: 5.4.0-53-generic
OS Type: 64-bit
Processors: 4 × Intel® Core™ i5-6200U CPU @ 2.30GHz
Memory: 31.2 ГиБ of RAM
Graphics Processor: Mesa Intel® HD Graphics 520

**Comment by elacuesta** [comment]
At first I thought this was probably related to #4855 (fixed by #4872), but I tried and I'm not even able to reproduce the issue on macOS:

```python
from scrapy import signals, Spider


class AsyncOpenExt:

    @classmethod
    def from_crawler(cls, crawler):
        ext = cls()
        crawler.signals.connect(ext.spider_opened, signal=signals.spider_opened)
        return ext

    async def spider_opened(self, spider: Spider):
        print("async spider_opened called")


class QuotesSpider(Spider):
    name = "quotes"
    start_urls = ["http://quotes.toscrape.com/tag/humor/"]
    custom_settings = {"EXTENSIONS": {AsyncOpenExt: 10}}

    def parse(self, response):
        for quote in response.css("div.quote"):
            yield {
                "author": quote.xpath("span/small/text()").get(),
                "text": quote.css("span.text::text").get(),
            }

```

```
scrapy runspider test-spiders/async_quotes.py -s TWISTED_REACTOR=twisted.internet.asyncioreactor.AsyncioSelectorReactor -s ASYNCIO_EVENT_LOOP=uvloop.Loop
```

<details>
  <summary>Full logs</summary>

  ```
2020-11-16 00:37:49 [scrapy.utils.log] INFO: Scrapy 2.4.0 started (bot: scrapybot)
2020-11-16 00:37:49 [scrapy.utils.log] INFO: Versions: lxml 4.5.2.0, libxml2 2.9.10, cssselect 1.1.0, parsel 1.6.0, w3lib 1.22.0, Twisted 20.3.0, Python 3.7.9 (v3.7.9:13c94747c7, Aug 15 2020, 01:31:08) - [Clang 6.0 (clang-600.0.57)], pyOpenSSL 19.1.0 (OpenSSL 1.1.1h  22 Sep 2020), cryptography 3.1.1, Platform Darwin-19.6.0-x86_64-i386-64bit
2020-11-16 00:37:49 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.asyncioreactor.AsyncioSelectorReactor
2020-11-16 00:37:49 [scrapy.utils.log] DEBUG: Using asyncio event loop: uvloop.Loop
2020-11-16 00:37:49 [scrapy.crawler] INFO: Overridden settings:
{'ASYNCIO_EVENT_LOOP': 'uvloop.Loop',
 'EDITOR': 'nano',
 'LOG_LEVEL': 'INFO',
 'SPIDER_LOADER_WARN_ONLY': True,
 'TWISTED_REACTOR': 'twisted.internet.asyncioreactor.AsyncioSelectorReactor'}
2020-11-16 00:37:49 [scrapy.extensions.telnet] INFO: Telnet Password: 4dc470409f64abb1
2020-11-16 00:37:49 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.memusage.MemoryUsage',
 'scrapy.extensions.logstats.LogStats',
 <class 'async_quotes.AsyncOpenExt'>]
2020-11-16 00:37:49 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2020-11-16 00:37:49 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.spidermiddlewares.referer.RefererMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
2020-11-16 00:37:49 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2020-11-16 00:37:49 [scrapy.core.engine] INFO: Spider opened
2020-11-16 00:37:49 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
async spider_opened called
2020-11-16 00:37:49 [scrapy.extensions.telnet] INFO: Telnet console listening on 127.0.0.1:6023
2020-11-16 00:37:50 [scrapy.core.engine] INFO: Closing spider (finished)
2020-11-16 00:37:50 [scrapy.statscollectors] INFO: Dumping Scrapy stats:
{'downloader/request_bytes': 229,
 'downloader/request_count': 1,
 'downloader/request_method_count/GET': 1,
 'downloader/response_bytes': 2214,
 'downloader/response_count': 1,
 'downloader/response_status_count/200': 1,
 'elapsed_time_seconds': 0.507076,
 'finish_reason': 'finished',
 'finish_time': datetime.datetime(2020, 11, 16, 3, 37, 50, 126022),
 'item_scraped_count': 10,
 'log_count/INFO': 10,
 'memusage/max': 53792768,
 'memusage/startup': 53792768,
 'response_received_count': 1,
 'scheduler/dequeued': 1,
 'scheduler/dequeued/memory': 1,
 'scheduler/enqueued': 1,
 'scheduler/enqueued/memory': 1,
 'start_time': datetime.datetime(2020, 11, 16, 3, 37, 49, 618946)}
2020-11-16 00:37:50 [scrapy.core.engine] INFO: Spider closed (finished)
  ```
</details>

**Versions**
```
Scrapy       : 2.4.0
lxml         : 4.5.2.0
libxml2      : 2.9.10
cssselect    : 1.1.0
parsel       : 1.6.0
w3lib        : 1.22.0
Twisted      : 20.3.0
Python       : 3.7.9 (v3.7.9:13c94747c7, Aug 15 2020, 01:31:08) - [Clang 6.0 (clang-600.0.57)]
pyOpenSSL    : 19.1.0 (OpenSSL 1.1.1h  22 Sep 2020)
cryptography : 3.1.1
Platform     : Darwin-19.6.0-x86_64-i386-64bit
```

**Comment by elacuesta** [comment]
Scratch that, I was working on an outdated branch 🤦 
The issue does happen in c340e72988fc6ec615b7b9851c3d28c16c26a839, but it appears after the process hangs (indeed because of #4855) and one tries to stop it with `ctrl+c`.
However, if you cherry pick the fix from #4872 (27b07c692c1bf9d35164a1cd8f8c81969a6d9e72), the spider does function as expected.

In short, closing as fixed by #4872.
